posix/dirent.c

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/dirent.h>
+#include <openenclave/corelibc/fcntl.h>
+#include <openenclave/corelibc/limits.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/trace.h>
+
+#define DIR_MAGIC 0x09180827
+
+struct _OE_DIR
+{
+    uint32_t magic;
+    int fd;
+    struct oe_dirent buf;
+};
+
+OE_DIR* oe_opendir_d(uint64_t devid, const char* pathname)
+{
+    OE_DIR* ret = NULL;
+    OE_DIR* dir = oe_calloc(1, sizeof(OE_DIR));
+    int fd = -1;
+
+    if (!dir)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if ((fd = oe_open_d(devid, pathname, OE_O_RDONLY | OE_O_DIRECTORY, 0)) < 0)
+        OE_RAISE_ERRNO_MSG(oe_errno, "pathname=%s", pathname);
+
+    dir->magic = DIR_MAGIC;
+    dir->fd = fd;
+
+    ret = dir;
+    dir = NULL;
+    fd = -1;
+
+done:
+
+    if (dir)
```

Comment:
should not need line 44, oe_free should deal with null like the standard free() does.

Reply:
I agree but this is our convention pretty much throughout the code. If we change this one, we should change hundreds of these across the OE source. If we feel stronly about this, then we can create an issue to clean these up.

Reply:
I'm a fan of the lazy cleanup approach.   That is, if you touch a function or a file, you clean up that part, and don't introduce new cases.   Requiring a global cleanup to be done all at once is not needed for things that are just perf, in my opinion.